### PR TITLE
tests: Bluetooth: Add babblesim tests for CSIP

### LINKED ISSUE #51380: LE Audio: CSIP invalid bsim tests
**Is your enhancement proposal related to a problem? Please describe.**
The current babblesim tests for CSIP mainly focuses on the case where everything goes right, and do not test much invalid behavior or error handling. 

**Describe the solution you'd like**
All APIs should be tested with invalid values to verify that they will return an error. 

To the extent possible, the APIs should also be used in cases where the the remote device should perform some error handling. 

Finally the APIs should be tested with disconnects. 

**Describe alternatives you've considered**
To test against a non-conforming device, some verification and checks in the stack needs to be disabled. TBD if this is a good idea to test somehow. 

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by szymon-czapracki**
Add tests for CSIP
Disable SIRK overwrite
Fix CHECKIF warnings

Don't know is this really needed right now after previously added tests for no size/rank.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/51380

### CODE REVIEW

File: include/zephyr/sys/check.h [hunk 1]
```diff
@@ -15,7 +15,8 @@
 	__ASSERT_NO_MSG(!(expr));   \
 	if (0)
 #elif defined(CONFIG_NO_RUNTIME_CHECKS)
-#define CHECKIF(...) \
+#define CHECKIF(expr) \
+	void(expr); \
```
Comment (Thalley) [inline include/zephyr/sys/check.h:19]: This change should be a separate commit

File: subsys/bluetooth/audio/csip_set_member.c [hunk 2]
```diff
@@ -808,19 +808,8 @@ int bt_csip_set_member_register(const struct bt_csip_set_member_register_param *
...
-
-		(void)memcpy(inst->set_sirk.value, test_sirk,
-			     sizeof(test_sirk));
-		LOG_DBG("CSIP SIRK was overwritten by sample data SIRK");
-	} else {
-		(void)memcpy(inst->set_sirk.value, param->set_sirk,
-			     sizeof(inst->set_sirk.value));
-	}
+	(void)memcpy(inst->set_sirk.value, param->set_sirk,
+		     sizeof(inst->set_sirk.value));
```
Comment (Thalley) [inline subsys/bluetooth/audio/csip_set_member.c:812]: Please update your commit message to explain why this is changed. Ideally commit messages should focus on why something was changed.

File: tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c [hunk 3]
```diff
@@ -207,10 +209,15 @@ static struct bt_le_scan_cb csip_set_coordinator_scan_callbacks = {
...
-	if (inst->info.set_size > 0) {
-		FAIL("Could not find all members (%u / %u)\n", members_found, inst->info.set_size);
+	if (expect_valid_sirk == false) {
+		PASS("Could not find all members (%u / %u)\n", members_found, inst->info.set_size);
 	} else {
+		if (inst->info.set_size > 0) {
+			FAIL("Could not find all members (%u / %u)\n", members_found,
+			     inst->info.set_size);
+		} else {
 		discover_timed_out = true;
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c:219]: Indentation missing

File: tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c [hunk 4]
```diff
@@ -235,6 +242,8 @@ static void ordered_access(const struct bt_csip_set_coordinator_set_member **mem
 		FAIL("Failed to do CSIP set coordinator ordered access (%d)",
 		      err);
 		return;
+	} else {
+		PASS("Failed to do CSIP set coordinator ordered access (%d)\n", err);
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c:246]: What's going on here? Is it supposed to fail or pass?

File: tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c [hunk 5]
```diff
@@ -244,6 +253,318 @@ static void ordered_access(const struct bt_csip_set_coordinator_set_member **mem
 	}
 }
 
+static void test_disc(void)
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c:256]:
```suggestion
static void test_discover(void)
```

File: tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c [hunk 6]
```diff
@@ -244,6 +253,318 @@ static void ordered_access(const struct bt_csip_set_coordinator_set_member **mem
...
+			FAIL("Failed to connect to %s: %d\n", addr, err);
+			return;
+		}
+
+		printk("Connected to %s\n", addr);
+		WAIT_FOR_FLAG(flag_connected);
+		connected_member_count++;
+
+		discovered = false;
+		printk("Doing discovery on member[%u]", i);
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c:360]:
```suggestion
		printk("Doing discovery on member[%u]\n", i);
```

File: tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c [hunk 7]
```diff
@@ -244,6 +253,318 @@ static void ordered_access(const struct bt_csip_set_coordinator_set_member **mem
...
+	bt_conn_disconnect(conns[0], BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+	printk("Member disconnected");
+
+	if (inst->info.lockable) {
+		printk("Locking set\n");
+		/* This should fail */
+		err = bt_csip_set_coordinator_lock(locked_members, connected_member_count,
+						   &inst->info);
+		if (err != 0) {
+			PASS("Failed to do set coordinator lock (%d)\n", err);
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c:389]: Confusing w.r.t. PASS and "failed"

Comment (Thalley) [reply tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c:389]: We also should not have more than 1 `PASS` for a test

File: tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c [hunk 8]
```diff
@@ -244,6 +253,318 @@ static void ordered_access(const struct bt_csip_set_coordinator_set_member **mem
...
+				printk("Disconnecting member[%u] (%s)", i, addr);
+				err = bt_conn_disconnect(conns[i],
+							 BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+				(void)memset(&set_members[i], 0, sizeof(set_members[i]));
+
+				return;
+			}
+		}
+	}
+	PASS("All members disconnected\n");
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c:400]:
```suggestion
	}
	
	PASS("All members disconnected\n");
```

File: tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c [hunk 9]
```diff
@@ -244,6 +253,318 @@ static void ordered_access(const struct bt_csip_set_coordinator_set_member **mem
...
+	connected_member_count++;
+
+	err = bt_csip_set_coordinator_discover(conns[0]);
+	if (err != 0) {
+		FAIL("Failed to initialize set coordinator for connection %d\n",
+		     err);
+		return;
+	}
+
+	WAIT_FOR_COND(discovered);
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c:459]: A lot of this can be moved to a common function for the tests

Comment (Thalley) [reply tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c:459]: A lot of the code below seems like it can also be moved to some common functions to reduce code duplication

File: tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c [hunk 10]
```diff
@@ -244,6 +253,318 @@ static void ordered_access(const struct bt_csip_set_coordinator_set_member **mem
...
+
+	/* Perform lock again */
+	err = bt_csip_set_coordinator_lock(locked_members, connected_member_count,
+					   &inst->info);
+	if (err != 0) {
+		FAIL("Failed to do set coordinator lock (%d)\n", err);
+	}
+
+	for (uint8_t i = 0; i < members_found; i++) {
+		printk("Disconnecting member[%u] (%s)", i, addr);
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c:554]:
```suggestion
		printk("Disconnecting member[%u] (%s)\n", i, addr);
```

File: tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c [hunk 11]
```diff
@@ -244,6 +253,318 @@ static void ordered_access(const struct bt_csip_set_coordinator_set_member **mem
...
+		if (err != 0) {
+			FAIL("Failed to do disconnect\n", err);
+			return;
+		}
+	}
+
+	PASS("All members disconnected\n");
+}
+
+
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c:567]:
```suggestion
}

```

File: tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c:1]: We haven't added much testing of invalid parameters for the individual functions, as requested by the issue https://github.com/zephyrproject-rtos/zephyr/issues/51380. 

Is this PR supposed to fix that issue, or is it a precursor to a PR that fixes the issue? If the former, the issue should not be considered in-review yet as significant amount of testing is missing (and the PR is missing the "fixes" keyword). If the latter, the issue should still not be considered in-review.

File: tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c [hunk 12]
```diff
@@ -445,6 +766,94 @@ static void test_main(void)
...
+	}
+
+	printk("Scanning successfully started\n");
+
+	WAIT_FOR_COND(members_found == 1);
+
+	printk("Stopping scan\n");
+	err = bt_le_scan_stop();
+	if (err != 0) {
+		FAIL("Could not stop scan");
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c:802]:
```suggestion
		FAIL("Could not stop scan\n");
```

File: tests/bsim/bluetooth/audio/src/csip_set_member_test.c [hunk 13]
```diff
@@ -22,6 +24,14 @@ struct bt_csip_set_member_register_param param = {
 		      0x22, 0xfd, 0xa1, 0x21, 0x09, 0x7d, 0x7d, 0x45 },
 };
 
+struct bt_csip_set_member_register_param param_unranked = {
+	.set_size = 3,
+	.lockable = true,
+	/* Using the CSIS test sample SIRK */
+	.set_sirk = { 0xcd, 0xcc, 0x72, 0xdd, 0x86, 0x8c, 0xcd, 0xce,
+		      0x22, 0xfd, 0xa1, 0x21, 0x09, 0x7d, 0x7d, 0x45 },
+};
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_member_test.c:33]: Any reason to have a duplicate param, rather than updating the `rank` value of the existing one?

File: tests/bsim/bluetooth/audio/src/csip_set_member_test.c [hunk 14]
```diff
@@ -105,6 +118,83 @@ static void test_main(void)
 	bt_conn_cb_register(&conn_callbacks);
 }
 
+static void test_disc(void)
+{
+	int err;
+
+	err = bt_enable(bt_ready);
+
+	if (err != 0) {
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_member_test.c:127]:
```suggestion
	err = bt_enable(bt_ready);
	if (err != 0) {
```

File: tests/bsim/bluetooth/audio/src/csip_set_member_test.c [hunk 15]
```diff
@@ -105,6 +118,83 @@ static void test_main(void)
...
+		FAIL("Bluetooth init failed (err %d)\n", err);
+		return;
+	}
+
+	bt_conn_cb_register(&conn_callbacks);
+
+	PASS();
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_member_test.c:134]: Missing pass text

File: tests/bsim/bluetooth/audio/src/csip_set_member_test.c [hunk 16]
```diff
@@ -105,6 +118,83 @@ static void test_main(void)
...
+	uint8_t rsi[BT_CSIP_RSI_SIZE];
+
+	struct bt_data ad[] = {
+		BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+		BT_CSIP_DATA_RSI(rsi),
+	};
+
+	err = bt_enable(bt_ready);
+
+	if (err != 0) {
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_member_test.c:149]:
```suggestion
	err = bt_enable(bt_ready);
	if (err != 0) {
```

File: tests/bsim/bluetooth/audio/src/csip_set_member_test.c [hunk 17]
```diff
@@ -105,6 +118,83 @@ static void test_main(void)
...
+
+	PASS();
+}
+
+static void test_reconnect(void)
+{
+	int err;
+	uint8_t rsi[BT_CSIP_RSI_SIZE];
+
+	struct bt_data ad[] = {
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_member_test.c:142]:
```suggestion
	uint8_t rsi[BT_CSIP_RSI_SIZE];
	struct bt_data ad[] = {
```

File: tests/bsim/bluetooth/audio/src/csip_set_member_test.c [hunk 18]
```diff
@@ -105,6 +118,83 @@ static void test_main(void)
...
+	bt_conn_cb_register(&conn_callbacks);
+
+	WAIT_FOR_COND(disc);
+
+	err = bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
+	if (err != 0) {
+		FAIL("Advertising failed to start (err %d)\n", err);
+	}
+
+	PASS();
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_member_test.c:163]: Missing PASS text

File: tests/bsim/bluetooth/audio/src/csip_set_member_test.c [hunk 19]
```diff
@@ -105,6 +118,83 @@ static void test_main(void)
...
+	err = bt_enable(bt_ready);
+
+	if (err != 0) {
+		FAIL("Bluetooth init failed (err %d)\n", err);
+		return;
+	}
+
+	bt_conn_cb_register(&conn_callbacks);
+
+	PASS();
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_member_test.c:179]: Missing PASS text

File: tests/bsim/bluetooth/audio/src/csip_set_member_test.c [hunk 20]
```diff
@@ -105,6 +118,83 @@ static void test_main(void)
...
+}
+
+static void test_unranked(void)
+{
+	int err;
+
+	err = bt_enable(bt_ready);
+
+	if (err != 0) {
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_member_test.c:172]:
```suggestion
	err = bt_enable(bt_ready);
	if (err != 0) {
```

File: tests/bsim/bluetooth/audio/src/csip_set_member_test.c [hunk 21]
```diff
@@ -105,6 +118,83 @@ static void test_main(void)
...
+	PASS();
+}
+
+static void test_sirk(void)
+{
+	int err;
+
+	err = bt_enable(bt_ready);
+
+	if (err != 0) {
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_member_test.c:188]:
```suggestion
	err = bt_enable(bt_ready);
	if (err != 0) {
```

File: tests/bsim/bluetooth/audio/src/csip_set_member_test.c [hunk 22]
```diff
@@ -105,6 +118,83 @@ static void test_main(void)
...
+		FAIL("Bluetooth init failed (err %d)\n", err);
+		return;
+	}
+
+	bt_conn_cb_register(&conn_callbacks);
+
+	PASS();
+}
+
+static void test_sirk(void)
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_member_test.c:182]: `test_sirk`, `test_unranked` and `test_disc` never start advertising, so how can they ever pass?

File: tests/bsim/bluetooth/audio/src/csip_set_member_test.c [hunk 23]
```diff
@@ -105,6 +118,83 @@ static void test_main(void)
...
+	err = bt_enable(bt_ready);
+
+	if (err != 0) {
+		FAIL("Bluetooth init failed (err %d)\n", err);
+		return;
+	}
+
+	bt_conn_cb_register(&conn_callbacks);
+
+	PASS();
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/src/csip_set_member_test.c:195]: Missing pass text

File: tests/bsim/bluetooth/audio/test_scripts/csip_disconnect.sh [hunk 24]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) 2023 Codecoup
+#
+# SPDX-License-Identifier: Apache-2.0
+
+# CSIP test where there is no rank and no lock characteristic on the Set Members
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/test_scripts/csip_disconnect.sh:7]: Copy pasted? :)

File: tests/bsim/bluetooth/audio/test_scripts/csip_invalid_sirk.sh [hunk 25]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) 2023 Codecoup
+#
+# SPDX-License-Identifier: Apache-2.0
+
+# CSIP test where there is no rank and no lock characteristic on the Set Members
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/test_scripts/csip_invalid_sirk.sh:7]: Copy pasted? :)

File: tests/bsim/bluetooth/audio/test_scripts/csip_disconnect.sh [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+
+source ${ZEPHYR_BASE}/tests/bsim/sh_common.source
+
+VERBOSITY_LEVEL=2
+EXECUTE_TIMEOUT=20
+
+cd ${BSIM_OUT_PATH}/bin
+
+SIMULATION_ID="csip_no_rank"
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/test_scripts/csip_disconnect.sh:16]: Copy pasted? :)

File: tests/bsim/bluetooth/audio/test_scripts/csip_invalid_sirk.sh [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+
+source ${ZEPHYR_BASE}/tests/bsim/sh_common.source
+
+VERBOSITY_LEVEL=2
+EXECUTE_TIMEOUT=20
+
+cd ${BSIM_OUT_PATH}/bin
+
+SIMULATION_ID="csip_no_rank"
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/test_scripts/csip_invalid_sirk.sh:16]: Copy pasted? :)

File: tests/bsim/bluetooth/audio/test_scripts/csip_reconnect.sh [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+# CSIP test where there is no rank and no lock characteristic on the Set Members
+
+source ${ZEPHYR_BASE}/tests/bsim/sh_common.source
+
+VERBOSITY_LEVEL=2
+EXECUTE_TIMEOUT=20
+
+cd ${BSIM_OUT_PATH}/bin
+
+SIMULATION_ID="csip_no_rank"
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/test_scripts/csip_reconnect.sh:16]: Copy pasted? :)

File: tests/bsim/bluetooth/audio/test_scripts/csip_reconnect.sh [hunk 29]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) 2023 Codecoup
+#
+# SPDX-License-Identifier: Apache-2.0
+
+# CSIP test where there is no rank and no lock characteristic on the Set Members
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/test_scripts/csip_reconnect.sh:7]: Copy pasted? :)

**Review by Thalley — CHANGES_REQUESTED** [review]

File: tests/bsim/bluetooth/audio/src/csip_set_member_test.c (hunk 22 above)
Comment (szymon-czapracki) [reply tests/bsim/bluetooth/audio/src/csip_set_member_test.c:182]: `bt_enable(bt_ready)` has advertising start within.

Comment (Thalley) [reply tests/bsim/bluetooth/audio/src/csip_set_member_test.c:182]: Ah, right, so it's in the callback - My bad :)

**Comment by Thalley** [comment]
@szymon-czapracki Any progress on this? Is it ready for review?
